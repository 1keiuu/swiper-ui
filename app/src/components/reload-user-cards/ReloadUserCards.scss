.reload-user__container {
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
  p {
    font-weight: bold;
    text-align: center;
    margin: 0;
    line-height: 40px;
    font-size: 17px;
  }
}

.empty-user__container {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  img {
    width: 70%;
  }
  p {
    font-weight: bold;
    text-align: center;
    margin: 0;
    line-height: 40px;
    font-size: 17px;
  }
}
