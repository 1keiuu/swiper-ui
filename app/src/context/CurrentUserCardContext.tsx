// NOTE: 現在表示中のcardの状態
import React, { createContext, useContext, useReducer } from "react";

type CurrentUserCardAction = {
  type: string;
  payload?: {
    status?: CardStatus;
    isFlipped?: boolean;
  };
};
type CurrentUserCardState = {
  index: number;
  status: CardStatus;
  isFlipped: boolean;
};
type CurrentUserCardDispatch = {
  incrementIndex: () => void;
  decrementIndex: () => void;
  changeStatus: (status: CardStatus) => void;
  setIsFlipped: (isFlipped: boolean) => void;
  toggleIsFlipped: () => void;
};

const initialState: CurrentUserCardState = {
  index: 0,
  status: "",
  isFlipped: false,
};

const CurrentUserCardStateContext =
  createContext<CurrentUserCardState>(initialState);
const CurrentUserCardDispatchContext = createContext<CurrentUserCardDispatch>({
  incrementIndex: () => {
    // do nothing.
  },
  decrementIndex: () => {
    // do nothing.
  },
  changeStatus: () => {
    // do nothing.
  },
  setIsFlipped: () => {
    // do nothing.
  },
  toggleIsFlipped: () => {
    // do nothing.
  },
});

export const useCurrentUserCardState = (): CurrentUserCardState =>
  useContext(CurrentUserCardStateContext);

export const useCurrentUserCardDispatch = (): CurrentUserCardDispatch =>
  useContext(CurrentUserCardDispatchContext);

const reducer = (
  state: CurrentUserCardState,
  action: CurrentUserCardAction
) => {
  const { payload } = action;
  switch (action.type) {
    case "INCREMENT_INDEX":
      return { ...state, index: state.index + 1 };
    case "DECREMENT_INDEX":
      return { ...state, index: state.index - 1 };
    case "CHANGE_STATUS":
      return { ...state, status: payload.status };
    case "SET_IS_FLIPPED":
      return { ...state, isFlipped: payload.isFlipped };
    case "TOGGLE_IS_FLIPPED":
      return { ...state, isFlipped: !state.isFlipped };
    default:
      throw new Error();
  }
};

export const CurrentUserCardProvider = ({
  children,
}: {
  children: JSX.Element;
}): JSX.Element => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const incrementIndex = () => {
    dispatch({ type: "INCREMENT_INDEX" });
  };
  const decrementIndex = () => {
    dispatch({ type: "DECREMENT_INDEX" });
  };
  const changeStatus = (status: CardStatus) => {
    dispatch({ type: "CHANGE_STATUS", payload: { status } });
  };
  const setIsFlipped = (isFlipped: boolean) => {
    dispatch({ type: "SET_IS_FLIPPED", payload: { isFlipped } });
  };
  const toggleIsFlipped = () => {
    dispatch({ type: "TOGGLE_IS_FLIPPED" });
  };

  return (
    <CurrentUserCardStateContext.Provider value={state}>
      <CurrentUserCardDispatchContext.Provider
        value={{
          incrementIndex,
          decrementIndex,
          changeStatus,
          setIsFlipped,
          toggleIsFlipped,
        }}
      >
        {children}
      </CurrentUserCardDispatchContext.Provider>
    </CurrentUserCardStateContext.Provider>
  );
};
