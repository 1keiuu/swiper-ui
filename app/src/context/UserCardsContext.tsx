import React, { createContext, useContext, useReducer } from "react";

type UserCardsAction = {
  type: string;
  payload?: {
    cards?: UserCard[];
    pageStatus?: PageStatus;
  };
};
type UserCardsState = {
  userCards: UserCard[];
  paginationIndex: number;
  pageStatus: PageStatus;
};
type UserCardsDispactch = {
  appendCards: (cards: UserCard[]) => void;
  resetCards: () => void;
  incrementPaginationIndex: () => void;
  setPageStatus: (pageStatus: PageStatus) => void;
};

const initialState: UserCardsState = {
  userCards: [],
  paginationIndex: 0,
  pageStatus: "reloading",
};

const UserCardsStateContext = createContext<UserCardsState>(initialState);

const UserCardsDispatchContext = createContext<UserCardsDispactch>({
  appendCards: () => {},
  resetCards: () => {},
  incrementPaginationIndex: () => {},
  setPageStatus: () => {},
});

export const useUserCardsState = (): UserCardsState => useContext(UserCardsStateContext);
export const useUserCardsDispatch = (): UserCardsDispactch => useContext(UserCardsDispatchContext);

const reducer = (state: UserCardsState, action: UserCardsAction) => {
  const { payload } = action;
  switch (action.type) {
    case "APPEND_CARDS":
      const sum = state.userCards.concat(payload.cards);
      return { ...state, userCards: sum };
    case "RESET_CARDS":
      return { ...state, userCards: [] };
    case "INCREMENT_PAGINATION_INDEX":
      return { ...state, paginationIndex: state.paginationIndex + 1 };
    case "SET_PAGE_STATUS":
      return { ...state, pageStatus: payload.pageStatus };
    default:
      throw new Error();
  }
};

export const UserCardsProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const appendCards = (cards: UserCard[]) => {
    dispatch({
      type: "APPEND_CARDS",
      payload: { cards },
    });
  };
  const resetCards = () => {
    dispatch({ type: "RESET_CARDS" });
  };
  const incrementPaginationIndex = () => {
    dispatch({ type: "INCREMENT_PAGINATION_INDEX" });
  };
  const setPageStatus = (pageStatus: PageStatus) => {
    dispatch({ type: "SET_PAGE_STATUS", payload: { pageStatus } });
  };

  return (
    <UserCardsStateContext.Provider value={state}>
      <UserCardsDispatchContext.Provider
        value={{
          appendCards,
          resetCards,
          incrementPaginationIndex,
          setPageStatus,
        }}
      >
        {children}
      </UserCardsDispatchContext.Provider>
    </UserCardsStateContext.Provider>
  );
};
