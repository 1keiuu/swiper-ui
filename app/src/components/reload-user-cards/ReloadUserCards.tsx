import React, { useContext, useEffect, useState } from "react";
import "./ReloadUserCards.scss";
import { CurrentCardStore } from "../../store/CurrentCard";
import fetch from "../../lib/fetch";
import { UserStore } from "../../store/User";

const ReloadUserCards: React.FC = () => {
  const { state, dispatch } = useContext(UserStore);
  const [status, setStatus] = useState("reloading");

  useEffect(() => {
    fetch((res) => {
      if (res.length == 0) setStatus("empty");
      dispatch({ type: "APPEND_USERS", incomingUsers: res });
    });
  }, []);
  if (status == "reloading") return <div>取得中...</div>;
  else if (status == "empty") return <div>empty</div>;
};

export default ReloadUserCards;
