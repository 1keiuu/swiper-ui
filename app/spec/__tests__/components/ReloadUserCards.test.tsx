import React from "react";
import { cleanup, render } from "@testing-library/react";
// FIXME: axiosがjest上で使えないので仮のコンポーネントを使ってテストしている
import ReloadUserCards from "../ignore/ReloadUserCards";

import * as UserCardsContext from "../../../src/context/UserCardsContext";

describe("ReloadUserCards Component", () => {
  // NOTE: Contextをスパイ化
  let userCardsStateSpy: jest.SpyInstance<unknown>;
  beforeEach(() => {
    userCardsStateSpy = jest.spyOn(UserCardsContext, "useUserCardsState");
    const initialState: UserCardsState = {
      userCards: [],
      paginationIndex: 0,
      pageStatus: "reloading",
    };
    userCardsStateSpy.mockImplementation(() => initialState);
  });
  afterEach(() => {
    userCardsStateSpy.mockClear();
    cleanup();
  });

  test("render reloading message when pageStatus is initial state.", () => {
    const { asFragment } = render(<ReloadUserCards />);
    expect(asFragment().firstChild.childNodes.length).toBe(1);
    expect(asFragment().firstChild.firstChild.textContent).toBe("取得中...");
    expect(asFragment()).toMatchSnapshot();
  });
  test("render empty message when pageStatus is 'empty'.", () => {
    const { asFragment, rerender } = render(<ReloadUserCards />);
    userCardsStateSpy.mockImplementation(() => ({
      userCards: [],
      paginationIndex: 0,
      pageStatus: "empty",
    }));
    // NOTE: userCardsStateSpyを使い、contextの値を変えて再描画
    rerender(<ReloadUserCards />);
    expect(asFragment().firstChild.childNodes.length).toBe(3);
    expect(asFragment()).toMatchSnapshot();
  });
});
