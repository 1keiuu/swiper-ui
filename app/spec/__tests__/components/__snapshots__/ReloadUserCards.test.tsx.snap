// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReloadUserCards Component render empty message when pageStatus is 'empty'. 1`] = `
<DocumentFragment>
  <div
    class="empty-user__container"
  >
    <p>
      スワイプできるカードがありません。
    </p>
    <p>
      時間を置いてから再度お試しください。
    </p>
    <img
      src="test-file-stub"
    />
  </div>
</DocumentFragment>
`;

exports[`ReloadUserCards Component render reloading message when pageStatus is initial state. 1`] = `
<DocumentFragment>
  <div
    class="reload-user__container"
  >
    <p>
      取得中...
    </p>
  </div>
</DocumentFragment>
`;
